src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -431,15 +437,34 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
       this.superinterfaces += superinterfaces
     }
 
-    fun addSuperinterface(superinterface: TypeName) = apply {
-      superinterfaces += superinterface
+    fun addSuperinterface(superinterface: TypeName, delegate: CodeBlock = CodeBlock.EMPTY) = apply {
+      if (delegate.isEmpty())
+        superinterfaces += superinterface
+      else {
+        require(kind == Kind.CLASS) {
+          "delegation only allowed for classes (found $kind '$name')" }
+        require(!superinterface.nullable) {
+          "expected non-nullable type but was '${superinterface.asNonNullable()}'"}
+        require(delegatedSuperInterfaces[superinterface] == null) {
+          "'$name' can not delegate to $superinterface by $delegate with existing declaration by " +
+              "${delegatedSuperInterfaces[superinterface]}" }
+        delegatedSuperInterfaces.put(superinterface, delegate)
+      }
     }
 
-    fun addSuperinterface(superinterface: Type)
-        = addSuperinterface(superinterface.asTypeName())
+    fun addSuperinterface(superinterface: Type, delegate: CodeBlock = CodeBlock.EMPTY)
+        = addSuperinterface(superinterface.asTypeName(), delegate)
+
+    fun addSuperinterface(superinterface: KClass<*>, delegate: CodeBlock = CodeBlock.EMPTY)
+        = addSuperinterface(superinterface.asTypeName(), delegate)
 
-    fun addSuperinterface(superinterface: KClass<*>)
-        = addSuperinterface(superinterface.asTypeName())
+    fun addSuperinterface(constructorParameterName: String) = apply {
+      requireNotNull(primaryConstructor) {
+        "delegating to constructor parameter requires not-null constructor" }
+      requireNotNull(primaryConstructor?.parameter(constructorParameterName)?.also {
+        param -> addSuperinterface(param.type, CodeBlock.of(param.name))
```

Comment:
The side-effecting use of `also` here seems really egregious. It's hard to follow this code. Can we split the validation and the effect into separate statements?